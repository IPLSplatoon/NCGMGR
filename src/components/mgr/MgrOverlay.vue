<template>
    <dialog
        ref="dialog"
        @close="onClose"
        @cancel.prevent="onCancel"
        @click.self="onClick"
    >
        <div class="content">
            <slot />
        </div>
    </dialog>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref, watch } from 'vue'

async function animationsComplete(element: HTMLElement): Promise<unknown> {
    return Promise.allSettled(element.getAnimations().map(animation => animation.finished))
}

export default defineComponent({
    name: 'MgrOverlay',

    props: {
        visible: {
            type: Boolean,
            required: true
        },
        noBackgroundClose: {
            type: Boolean,
            default: false
        },
        maxWidth: {
            type: Boolean,
            default: false
        }
    },

    emits: ['update:visible'],

    setup (props, { emit }) {
        const dialog = ref<HTMLDialogElement | null>(null)

        async function onClose() {
            if (!dialog.value) return
            dialog.value.setAttribute('inert', '')
            await animationsComplete(dialog.value)
            dialog.value.close()
        }

        onMounted(() => {
            if (!props.visible) {
                dialog.value?.setAttribute('inert', '')
            } else {
                dialog.value?.showModal()
            }

            watch(() => props.visible, newValue => {
                if (newValue) {
                    dialog.value?.removeAttribute('inert')
                    dialog.value?.showModal()
                } else {
                    onClose()
                }
            })
        })

        return {
            dialog,
            onClose() {
                emit('update:visible', false)
            },
            onCancel() {
                if (!props.noBackgroundClose) {
                    onClose()
                }
            },
            onClick() {
                if (!props.noBackgroundClose) {
                    onClose()
                }
            }
        }
    }
})
</script>

<style lang="scss" scoped>
@use '../../styles/constants';

dialog {
    border: none;
    background-color: var(--ipl-bg-primary);
    color: var(--ipl-body-text-color);
    width: 85%;
    max-height: 80%;
    height: max-content;
    box-sizing: border-box;
    //margin: 0;
    padding: 0;
    inset: 0;
    display: block;
    position: fixed;
    // Unfortunately, if you want animation, you'll have to put up with this.
    z-index: 999999;
    transform: scale(0.9);
    opacity: 0;
    border-radius: constants.$border-radius-outer;

    &:not([open]) {
        pointer-events: none;
    }

    &[open] {
        animation: enter constants.$transition-duration-low ease-out forwards;

        &::backdrop {
            animation: backdrop-in constants.$transition-duration-low ease-out forwards;
        }

        &[inert] {
            animation: exit constants.$transition-duration-low ease-in forwards;

            &::backdrop {
                animation: backdrop-out constants.$transition-duration-low ease-out forwards;
            }
        }
    }

    &::backdrop {
        background: var(--ipl-page-overlay-color);
    }

    > .content {
        padding: 8px;
        box-sizing: border-box;
    }

    &:focus-visible {
        outline: var(--ipl-focus-outline-color) solid var(--ipl-focus-outline-width);
    }
}

@keyframes enter {
    to {
        transform: scale(1);
        opacity: 1;
    }
}

@keyframes exit {
    0% {
        transform: scale(1);
        opacity: 1;
    }

    100% {
        transform: scale(0.9);
        opacity: 0;
    }
}

@keyframes backdrop-in {
    0% {
        opacity: 0;
    }

    100% {
        opacity: 1;
    }
}

@keyframes backdrop-out {
    0% {
        opacity: 1;
    }

    100% {
        opacity: 0;
    }
}
</style>
