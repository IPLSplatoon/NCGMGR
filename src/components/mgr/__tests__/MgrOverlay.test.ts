import MgrOverlay from '@/components/mgr/MgrOverlay.vue'
import { mount } from '@vue/test-utils'

describe('MgrOverlay', () => {
    it('matches snapshot when visible', () => {
        const wrapper = mount(MgrOverlay, {
            props: {
                visible: true
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('matches snapshot when visible and max width', () => {
        const wrapper = mount(MgrOverlay, {
            props: {
                visible: true,
                maxWidth: true
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('matches snapshot when hidden', () => {
        const wrapper = mount(MgrOverlay, {
            props: {
                visible: false
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('closes itself on background click if noBackgroundClose is false', () => {
        const wrapper = mount(MgrOverlay, {
            props: {
                visible: true,
                noBackgroundClose: false
            }
        })

        wrapper.get('.mgr-overlay__wrapper').trigger('click')

        const emitted = wrapper.emitted('update:visible')
        expect(emitted?.length).toEqual(1)
        expect(emitted?.[0]).toEqual([false])
    })

    it('does nothing on background click if noBackgroundClose is true', () => {
        const wrapper = mount(MgrOverlay, {
            props: {
                visible: true,
                noBackgroundClose: true
            }
        })

        wrapper.get('.mgr-overlay__wrapper').trigger('click')

        expect(wrapper.emitted('update:visible')).toBeFalsy()
    })
})
