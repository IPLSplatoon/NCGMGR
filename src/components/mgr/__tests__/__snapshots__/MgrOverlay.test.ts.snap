// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MgrOverlay matches snapshot when hidden 1`] = `
<transition-stub>
  <!--v-if-->
</transition-stub>
`;

exports[`MgrOverlay matches snapshot when visible 1`] = `
<transition-stub>
  <div class="mgr-overlay__wrapper layout horizontal center-horizontal center-vertical">
    <div class="mgr-overlay__content"></div>
  </div>
</transition-stub>
`;

exports[`MgrOverlay matches snapshot when visible and max width 1`] = `
<transition-stub>
  <div class="mgr-overlay__wrapper layout horizontal center-horizontal center-vertical">
    <div class="mgr-overlay__content max-width"></div>
  </div>
</transition-stub>
`;
